urlb54e207112_Ghuillaume_CSA_tgz-pJ8-M0_ApplicationJ8/src/M2/Configuration.java

```diff
@@ -0,0 +1,370 @@
+		}
+	}
+	
+	public final void addPort(PortConfigFourni pc) {
+		if(this.tags.contains(pc.getName())) {
+			Trace.printError("Tag " + pc.getName() + " is not available, service as not been created");
+		}
+		else {
+			this.portsFourni.put(pc.getName(), pc);
+			this.tags.add(pc.getName());
+		}
+	}
+	
+	public final void addPort(PortConfigRequis pc) {
+		if(this.tags.contains(pc.getName())) {
+			Trace.printError("Tag " + pc.getName() + " is not available, service as not been created");
+		}
+		else {
+			this.portsRequis.put(pc.getName(), pc);
+			this.tags.add(pc.getName());
+		}
+	}
+	
+	
+	// Getters
+
+	public final PortConfigRequis getPortR(String name) {
+		return this.portsRequis.get(name);
+	}
+	
+	public final PortConfigFourni getPortF(String name) {
+		return this.portsFourni.get(name);
+	}
+	
+	
+	public final Composant getComposant(String name) {
+		Iterator<Composant> it = this.components.iterator();
+		while(it.hasNext()) {
+			Composant currentComponent = it.next();
+			if(currentComponent.getName().equals(name)) {
+				return currentComponent;
+			}
+		}
+		
+		return null;
+	}
+	
+	public final Connecteur getConnector(String name) {
+		Iterator<Connecteur> it = this.connectors.iterator();
+		while(it.hasNext()) {
+			Connecteur currentConnector = it.next();
+			if(currentConnector.getName().equals(name)) {
+				return currentConnector;
+			}
+		}
+		
+		return null;
+	}
+	
+	public final Configuration getParentConfig() {
+		Configuration parentConfig = null;
+		
+		if(this.parent != null) {
+			if(this.parent.getParent() != null) {
+				if(this.parent.getParent() instanceof Configuration) {
+					parentConfig = (Configuration)this.parent.getParent();
+				}
+			}
+		}
+		
+		return parentConfig;
+	}
+	
+	
+	// Attachments (2)
+	
+	public final Attachment addLink(String name, RoleFrom r, PortFourni p) {
+		Attachment a = null;
+		if(this.tags.contains(name)) {
+			Trace.printError("Tag " + name + " is not available, service as not been created");
+		}
+		else {
+			if (r.getConfig().equals(this) && p.getConfig().equals(this)) {
+				a = new Attachment(name);
+				a.bind(r, p);
+				this.attachments.add(a);
+				this.tags.add(a.getName());
+			}
+		}
+		return a;
+	}
+	
+	public final Attachment addLink(String name, RoleTo r, PortRequis p) {
+		Attachment a = null;
+		if(this.tags.contains(name)) {
+			Trace.printError("Tag " + name + " is not available, service as not been created");
+		}
+		else {
+			a = new Attachment(name);
+			a.bind(r, p);
+			this.attachments.add(a);
+			this.tags.add(a.getName());
+		}
+		return a;
+	}
+	
+	
+	// Bindings (4)
+	
+	public final Binding addLink(String name, RoleFrom r, PortConfigFourni p) {
+		Binding b = null;
+
+		if(this.tags.contains(name)) {
+			Trace.printError("Tag " + name + " is not available, service as not been created");
+		}
+		else {
+			b = new Binding(name);
+			b.bind(r, p);
+			this.bindings.add(b);
+			this.tags.add(b.getName());
+		}
+		return b;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 194:205, 209:220, 246:257.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.